builtin/logical/pki/backend.go

```diff
@@ -145,31 +151,49 @@ the "config/" path.
 `
 
 func metricsKey(req *logical.Request, extra ...string) []string {
+	if req == nil || req.MountPoint == "" {
+		return extra
+	}
 	key := make([]string, len(extra)+1)
 	key[0] = req.MountPoint[:len(req.MountPoint)-1]
 	copy(key[1:], extra)
 	return key
 }
 
-func (b *backend) metricsWrap(callType string, ofunc roleOperation) framework.OperationFunc {
+func (b *backend) metricsWrap(callType string, roleMode int, ofunc roleOperation) framework.OperationFunc {
 	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
 		key := metricsKey(req, callType)
-		roleName := data.Get("role").(string)
-
-		// Get the role
-		role, err := b.getRole(ctx, req.Storage, roleName)
-		if err != nil {
-			return nil, err
+		var role *roleEntry
+		var labels []metrics.Label
+		var err error
+
+		var roleName string
+		switch roleMode {
+		case roleRequired:
+			roleName = data.Get("role").(string)
+		case roleOptional:
+			r, ok := data.GetOk("role")
+			if ok {
+				roleName = r.(string)
+			}
 		}
-		if role == nil {
-			return logical.ErrorResponse(fmt.Sprintf("unknown role: %s", roleName)), nil
+		if roleMode > noRole {
```

Comment:
I think this would be clearer with a != ;
Would also be cool to throw an error if the roleMode is not one of the possible values.